sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ReceiverManager.cs

```diff
@@ -0,0 +1,358 @@
+        }
+
+        /// <summary>
+        ///
+        /// </summary>
+        /// <param name="message"></param>
+        /// <param name="cancellationTokenSource"></param>
+        /// <returns></returns>
+        private async Task RenewMessageLock(
+        ServiceBusReceivedMessage message,
+        CancellationTokenSource cancellationTokenSource)
+        {
+            cancellationTokenSource.CancelAfter(_processorOptions.MaxAutoLockRenewalDuration);
+            CancellationToken cancellationToken = cancellationTokenSource.Token;
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                try
+                {
+                    ServiceBusEventSource.Log.ProcessorRenewMessageLockStart(_identifier, 1, message.LockToken);
+                    TimeSpan delay = CalculateRenewDelay(message.LockedUntil);
+
+                    try
+                    {
+                        await Task.Delay(delay, cancellationToken).ConfigureAwait(false);
+                    }
+                    catch (TaskCanceledException)
+                    {
+                        break;
+                    }
+                    if (Receiver.IsDisposed)
+                    {
+                        break;
+                    }
+
+                    await Receiver.RenewMessageLockAsync(message, cancellationToken).ConfigureAwait(false);
+                    ServiceBusEventSource.Log.ProcessorRenewMessageLockComplete(_identifier);
+                }
+                catch (Exception ex) when (!(ex is TaskCanceledException))
+                {
+                    ServiceBusEventSource.Log.ProcesserRenewMessageLockException(_identifier, ex);
+                    await HandleRenewLockException(ex, cancellationToken).ConfigureAwait(false);
+
+                    // if the error was not transient, break out of the loop
+                    if (!(ex as ServiceBusException)?.IsTransient == true)
+                    {
+                        break;
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Cancels the specified cancellation source and awaits the specified task.
+        /// </summary>
+        /// <param name="cancellationSource">CancellationTokenSource to cancel</param>
+        /// <param name="task">Associated task to await</param>
+        private static async Task CancelTask(
+                CancellationTokenSource cancellationSource,
+                Task task)
+        {
+            try
+            {
+                if (cancellationSource != null)
+                {
+                    cancellationSource.Cancel();
+                    await task.ConfigureAwait(false);
+                }
+            }
+            catch (Exception ex) when (ex is TaskCanceledException)
+            {
+                // Nothing to do here.  These exceptions are expected.
+            }
+        }
+
+        private void ThrowIfSessionLockLost(
```

Comment:
Interesting to see session related stuff in this file

Reply:
Yes, since ProcessOneMessage is not virtual and only in ReceiverManager. I think it is probably okay.